### PR TITLE
feat(file): alternate file implementation

### LINKED ISSUE #123: File moveFile function call doesn't trigger change detection
I am using this method to move a file around the disk.

```
File.moveFile("myPathWithoutFilename", "fileName", window.cordova.file.dataDirectory, "myNewFileName).then(fileDescriptor => {
   this.myImagePath = fileDescriptor.nativeURL;
});
```

Change detection is not run and the change to `this.myImagePath` is not recognized.

However, if I wrap the call to this
```
File.moveFile("myPathWithoutFilename", "fileName", window.cordova.file.dataDirectory, "myNewFileName).then(fileDescriptor => {
   this.ngZone.run( () => {
      this.myImagePath = fileDescriptor.nativeURL;
   });
});
```

Change detection does fire and the bindings update.

I am happy to provide a sample project if needed.

### LINKED ISSUE #159: File class is missing a writeFile() method
As far as I can tell there isn't a method for writing files using the `ionic-native` File class. In ngCordova [this method was available](http://ngcordova.com/docs/plugins/file/#writefilepath-file-data-replace) and with the actual File API you can use [this method](https://github.com/apache/cordova-plugin-file#write-to-a-file-).

Was this missed out for a reason? If not would it be possible to add something for this thanks.

### LINKED ISSUE #233: Add new implementation for File plugin
This issue is just for reference and as a reminder.

New file implementation is in the dev branch. Commit URL: https://github.com/driftyco/ionic-native/commit/b1caab316a2d8f2ef10d8b3c6054aba63437823d

---

- Add missing functionality
- Test
- List as a breaking change in next release

### LINKED ISSUE #254: File.listDir doesn't work on Android 4.4
Seems fine on newer devices / emulators, but my 4.4 test device and 4.4 emulator both fail. I get a promise catch with just the object { "code": "JSON error }

I am accessing the cordova plugin directly to get the cache directory. My call looks like this:
```
File.listDir(cordova.file.cacheDirectory, "slides").then(list => {
  //do stuff with list
}).catch(err => {
  //receives err = { "code": "JSON error } in Android 4.4
}
```

Happens on ionic-native 1.2.4 and 1.3.2

### LINKED ISSUE #264: Write File Support
In the `File` plugin, we should add a method for writing files.

It is kind of painful the way it is now.  It was very easy in [ngCordova](http://ngcordova.com/docs/plugins/file/). `writeFile(path, file, data, replace)`

I wrote this method to do it in a small sample app.  It works but we need additional rigor around it.

```
writeFileInternal(directoryPath, fileName, fileContent){
    return new Promise( (resolve, reject) => {
      (<any>window).resolveLocalFileSystemURL(directoryPath, (fileSystem) => {
        fileSystem.getFile(fileName, {create: true}, (fileEntry) => {
          fileEntry.createWriter( (writer) => {
            writer.onwriteend = function (evt) {
              if (this.error) {
                reject(new Error("Failed to write file"));
              } else {
                resolve(evt);
              }
            };

            writer.write(fileContent);

          }, (err) => {
            console.log("ERROR: ", err);
            reject(err);
          });
        });
      });
    });
```

Thanks,
Dan

### PR DESCRIPTION
**Pull request by ihadeed**
This is an alternative file implementation, written by @rapropos 

closes #264
closes #254 
closes #233
closes #159
closes #123

### CODE REVIEW

**Comment by ihadeed** [comment]
It would be appreciated if someone can give this a quick test: 
https://www.dropbox.com/s/x5589dwg93qdvap/ionic-native-1.3.2.tgz?dl=0

**Comment by ghenry22** [comment]
I'll try and give it a test this week

**Comment by rapropos** [comment]
Do I need to do something to address the conflict warning?

**Comment by ihadeed** [comment]
@rapropos 

I'll fix the conflict.

I was wondering, do all the functions in this implementation work a 100%? If so I will just merge it without testing it personally.

**Comment by rapropos** [comment]
Sorry, I have not tested every single function, because I don't have client code that uses all of them.

File: src/index.ts [hunk 1]
```diff
@@ -93,7 +93,11 @@ export * from './plugins/datepicker';
 export * from './plugins/device';
 export * from './plugins/devicemotion';
 export * from './plugins/deviceorientation';
+<<<<<<< HEAD
+export * from './plugins/file';
+=======
 export * from './plugins/facebook';
+>>>>>>> master
```
Comment (danbucholtz) [inline src/index.ts]: @ihadeed, looks like some git merge issues in this file.

Thanks,
Dan

Comment (ihadeed) [inline src/index.ts]: @danbucholtz Already fixed! 😁

**Comment by danbucholtz** [comment]
@ihadeed, you rock! 💯

**Comment by teledemic** [comment]
I can see the different directory constants (cacheDirectory etc) are in there which is great. But I can't figure out how to access them! Can you post an example of a simple "check if file exists" or similar?

**Pull request merged**